### PR TITLE
SDK-1301: Add DocScanClient

### PR DESCRIPTION
**Pull request by davidgrayston**
### Added
- `DocScanClient`
  - `.createSession()`
  - `.getSession()`
  - `.deleteSession()`
  - `.getMedia()`
  - `.deleteMedia()`
- Added `Media`, which can be created with `{ByteBuffer|Buffer}` _content_, and `{string}` _mimeType_
  > Note: `{ByteBuffer}` is the existing type being used for image content, which comes from `protobufjs` _(Changing this to `{Buffer}` only would be a breaking change)_.
  >
  > As a result, _bytebuffer_ has been added to `package.json`, but will likely be removed and replaced by `{Buffer}` in the next major release.

### Changed
- `Image` now extends `Media`
- Renamed `appId` parameters to `sdkId`

### Fixed
- `Validation.isNumber()` now correctly checks the type of values

### Deprecated
- `YotiClient.applicationId` replaced by `sdkId`
- `SandboxClientBuilder.forApplication()` replaced by `withClientSdkId()`

### CODE REVIEW

File: src/doc_scan_service/session/create/check/requested.liveness.check.builder.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  /**
+   * Sets the type of the liveness check to the supplied value
+   *
+   * @param {string} livenessType
+   *   The type of the liveness check
+   *
+   * @returns {this}
+   */
+  forLivenessType(livenessType) {
```
Comment (emmas-yoti) [inline src/doc_scan_service/session/create/check/requested.liveness.check.builder.js]: It feels kinda weird for these to be for?

Is there a good reason they're not with? We use with to set strings in other places.

Comment (davidgrayston) [reply src/doc_scan_service/session/create/check/requested.liveness.check.builder.js]: It's just following the Java implementation - thoughts @echarrod / @MrBurtyyy ?

Comment (MrBurtyyy) [reply src/doc_scan_service/session/create/check/requested.liveness.check.builder.js]: I think `for` makes sense, as the config the check is _for_ a zoom liveness check, _with_ a maximum number of retries right?

File: src/doc_scan_service/session/retrieve/doc.scan.session.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+const Validation = require('../../../yoti_common/validation');
+const ResourceContainer = require('./resource.container');
+const CheckResponse = require('./check.response');
+const AuthenticityCheckResponse = require('./authenticity.check.response');
+const FaceMatchCheckResponse = require('./face.match.check.response');
+const TextDataCheckResponse = require('./text.data.check.response');
+const ZoomLivenessCheckResponse = require('./zoom.liveness.check.response');
+const DocScanConstants = require('../../doc.scan.constants');
+
+class DocScanSession {
```
Comment (MrBurtyyy) [inline src/doc_scan_service/session/retrieve/doc.scan.session.js]: This should be called `GetSessionResult` to bring it in-line with already released Java SDK

File: src/doc_scan_service/session/retrieve/document.resource.response.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+const ResourceResponse = require('./resource.response');
+const DocumentFieldsResponse = require('./document.fields.response');
+const PageResponse = require('./page.response');
+const Validation = require('../../../yoti_common/validation');
+
+class DocumentResourceResponse extends ResourceResponse {
```
Comment (MrBurtyyy) [inline src/doc_scan_service/session/retrieve/document.resource.response.js]: This should be called `IdDocumentResourceResponse` to be inline with Java

File: src/doc_scan_service/session/retrieve/resource.container.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+const DocScanConstants = require('../../doc.scan.constants');
+const Validation = require('../../../yoti_common/validation');
+
+class ResourceContainer {
+  constructor(resources) {
+    if (resources.id_documents) {
+      Validation.isArray(resources.id_documents, 'id_documents');
+      this.idDocuments = resources
+        .id_documents
+        .map(resource => new DocumentResourceResponse(resource));
```
Comment (MrBurtyyy) [inline src/doc_scan_service/session/retrieve/resource.container.js]: This should be `new IdDocumentResourceResponse(resource)`

File: src/doc_scan_service/session/retrieve/task.response.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    this.id = task.id;
+
+    Validation.isString(task.state, 'state', true);
+    this.state = task.state;
+
+    Validation.isString(task.created, 'created', true);
+    this.created = task.created;
+
+    Validation.isString(task.last_updated, 'last_updated', true);
+    this.lastUpdated = task.last_updated;
```
Comment (MrBurtyyy) [inline src/doc_scan_service/session/retrieve/task.response.js]: These are set below, and so these lines can be removed (must be carry over where we decided on using dates)

File: src/doc_scan_service/session/retrieve/zoom.liveness.check.response.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+'use strict';
+
+const CheckResponse = require('./check.response');
+
+class ZoomLivenessCheckResponse extends CheckResponse {
+}
+
+module.exports = ZoomLivenessCheckResponse;
```
Comment (MrBurtyyy) [inline src/doc_scan_service/session/retrieve/zoom.liveness.check.response.js]: I don't think we have this in Java or PHP (it's called `LivenessCheckResponse`).  Seeing as Java has already been released, do you think we should keep it the same?

File: src/doc_scan_service/session/retrieve/doc.scan.session.js (hunk 2 above)
Comment (davidgrayston) [reply src/doc_scan_service/session/retrieve/doc.scan.session.js]: Renamed in 4966c52

File: src/doc_scan_service/session/retrieve/zoom.liveness.check.response.js (hunk 6 above)
Comment (davidgrayston) [reply src/doc_scan_service/session/retrieve/zoom.liveness.check.response.js]: Renamed in 30a4f39

File: src/doc_scan_service/session/retrieve/task.response.js (hunk 5 above)
Comment (davidgrayston) [reply src/doc_scan_service/session/retrieve/task.response.js]: Fixed in 51076f8

File: src/doc_scan_service/session/retrieve/resource.container.js (hunk 4 above)
Comment (davidgrayston) [reply src/doc_scan_service/session/retrieve/resource.container.js]: Renamed in d896ddd

File: src/doc_scan_service/doc.scan.service.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        .catch(err => reject(new DocScanError(err)));
+    });
+  }
+
+  /**
+   * Retrieves the current state of a given session
+   *
+   * @param {string} sessionId
+   *
+   * @returns {Promise} Resolves DocScanSession
```
Comment (MrBurtyyy) [inline src/doc_scan_service/doc.scan.service.js]: This should resolve `GetSessionResult`

Comment (davidgrayston) [reply src/doc_scan_service/doc.scan.service.js]: Fixed in 52cf4bd

File: src/client/doc.scan.client.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  constructor(sdkId, pem) {
+    this.docScanService = new DocScanService(sdkId, pem);
+  }
+
+  /**
+   * Creates a Doc Scan session using the supplied session specification
+   *
+   * @param {SessionSpecification} sessionSpecification
+   *
+   * @returns {Promise} Resolving SessionResult
```
Comment (MrBurtyyy) [inline src/client/doc.scan.client.js]: Should be `CreateSessionResult`

File: tests/client/doc.scan.client.spec.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+const uuid = require('uuid');
+
+const config = require('../../config');
+
+const {
+  DocScanClient,
+  SessionSpecificationBuilder,
+} = require('../../');
+
+const SessionResult = require('../../src/doc_scan_service/session/create/create.session.result');
```
Comment (MrBurtyyy) [inline tests/client/doc.scan.client.spec.js]: Should we change this to `CreateSessionResult` so it's named the same as the actual class you're requiring?

File: tests/doc_scan_service/doc.scan.service.spec.js [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+const uuid = require('uuid');
+
+const config = require('../../config');
+
+const {
+  SessionSpecificationBuilder,
+  DocScanService,
+} = require('../../src/doc_scan_service');
+
+const SessionResult = require('../../src/doc_scan_service/session/create/create.session.result');
```
Comment (MrBurtyyy) [inline tests/doc_scan_service/doc.scan.service.spec.js]: Same here

File: tests/client/doc.scan.client.spec.js (hunk 9 above)
Comment (davidgrayston) [reply tests/client/doc.scan.client.spec.js]: Fixed in 46e3c65

File: tests/doc_scan_service/doc.scan.service.spec.js (hunk 10 above)
Comment (davidgrayston) [reply tests/doc_scan_service/doc.scan.service.spec.js]: Fixed in 46e3c65

File: src/doc_scan_service/doc.scan.service.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+/**
+ * @param {string} sessionId
+ * @param {string} mediaId
+ *
+ * @returns {string}
+ */
+const mediaContentPath = (sessionId, mediaId) => `${sessionPath(sessionId)}/media/${mediaId}/content`;
+
+/**
+ * Service built to handle the interactions between the client and Doc Scan API's
```
Comment (echarrod) [inline src/doc_scan_service/doc.scan.service.js]:
```suggestion
 * Service built to handle the interactions between the client and Doc Scan APIs
```

File: src/doc_scan_service/doc.scan.service.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+          } catch (err) {
+            return reject(new DocScanError(err));
+          }
+        })
+        .catch(err => reject(new DocScanError(err)));
+    });
+  }
+
+  /**
+   * Deletes a session and all of it's associated content
```
Comment (echarrod) [inline src/doc_scan_service/doc.scan.service.js]:
```suggestion
   * Deletes a session and all of its associated content
```
I realised these two apostrophes are present in Java too. 
@MrBurtyyy no rush, but please could you update them in Java too?

Comment (davidgrayston) [reply src/doc_scan_service/doc.scan.service.js]: Corrected in 0b91cfd

File: src/doc_scan_service/doc.scan.service.js (hunk 11 above)
Comment (davidgrayston) [reply src/doc_scan_service/doc.scan.service.js]: Corrected in 0b91cfd

File: tests/dynamic_sharing_service/extension/location.constraint.extension.builder.spec.js [hunk 13]
```diff
@@ -23,24 +23,30 @@ describe('LocationConstraintExtensionBuilder', () => {
 
   it('should fail for longitude too low', () => {
     const builder = new LocationConstraintExtensionBuilder()
+      .withLatitude(0)
```
Comment (emmas-yoti) [inline tests/dynamic_sharing_service/extension/location.constraint.extension.builder.spec.js]: This isn't referenced in the changelog. Did it sneak in?

**Comment by davidgrayston** [comment]
@echarrod / @MrBurtyyy can you take a look at the last commit? bfd539f04f0eb568947685ad8d082187fa75de51

**Pull request merged**